### ISSUE TITLE
Instead of writing 0_master.hkx to the Overwrite after FNIS run, MO replaces Vilja/USLEEP mod file

### ISSUE THREAD

**Issue by REALfore** [TanninOne/modorganizer#647]
Both mods have the peculiarity that they include 0_master.hkx twice. Inside a bsa, AND as open file.

When FNIS is run, MO will overwrite the Vilja/USLEEP 0_master.hkx with the file generated by FNIS. When both these mods are active, MO will overwrite the mod with lower priority.

MO will not overwrite 0_master.hkx, when a mod includes this file only once (either in bsa, or as open file).

This is a very severe issue for me. Because without 0_master.hkx most FNIS based animation mods will not work at all. And there is no indication at all about an issue.

log file: https://drive.google.com/open?id=0B4wJfSHB28r8NldseEd4SndRUUE
________________

A little side note. I have an issue with FNIS under Win 10, where a file function (GetFileInfo) doesn't work properly any more, when the path contains a double "\\". Has worked with Win7 before. I only mention this, because I saw you have "\\" as well in your log file.

**Comment by GrantSP** [comment]
In discussion with @REALfore about this matter it can be shown that MO **does not treat loose files and BSA files the same way**, despite the 'Data' tab showing those files as available.

With a very basic Skyrim installation and a completely fresh MO install that contains only FNIS and the Vilja mod installed. It can be shown that there are 3 copies of the _0_master.hkx_ file.
One loose in the FNIS mod and a loose and a BSA copy in the Vilja mod.

Running FNIS with these mods ordered:
FNIS - priority 1
Vilja - priority 2
will result in the loose file from the Vilja mod being modified by FNIS and it will be found inside that mod.
This is expected as MO is said to: "modify existing files and save them into their respective mod folders and only create new files and save them into 'Overwrite'".
However in order to use the loose file from FNIS, if you simply delete the loose file from the Vilja mod the tool will error out saying the needed file cannot be found. This is despite the fact the file, _0_master.hkx_, is shown in the 'Data' tab as provided by both FNIS and the winning mod: Vilja (inside a BSA).

Only when the order of those mods is reversed is the file now said to be available and the FNIS tool runs properly.

So, although the title of this issue is incorrect, ie. the behaviour is expected, there is shown by this example the fact that files inside BSAs are not made available to all tools as though they are loose, which the 'Data' tab is implying.

The main description on the MO Nexus page says:

> BSA Priority (3):
> -----------------
> BSA files are archives that contain assets like textures, meshes, sounds, scripts, ...
> All vanilla data resides in BSAs and by default, the Skyrim Creation Kit packages assets for mods in BSAs. Traditionally you can not control priority between BSAs and "loose" files (that is: assets that came as individual files), BSAs are always overwritten. Therefore, if you wanted to use files from a BSA over files from a mod that came as "loose" files, you'd have to figure out and delete the correct files manually.
> 
> MO fixes this problems and afaik it is the only tool available to do so. In the UI, check all BSAs you want loaded and you're ready to go. Files from BSAs have priority according to their installation order (left pane) -> They overwrite files from mods with lower priorities and don't overwrite files with higher priority. The limit of enables BSAs is approx 8000)

This is shown to NOT be the case. Files from BSAs are not seen in the VFS even when there is NO CONFLICT.
Using the documented trick to examine the VFS, Free Commander XE, reveals that nothing inside the BSAs are provided.

**Comment by TanninOne** [comment]
You misunderstood the description. It says files from the bsa gain priority according to their installation order, it doesn't say it magically makes files from inside the archive visible to tools that don't support that archive format, much less allow you to overwrite them.

**Comment by GrantSP** [comment]
In the documentation, FAQ section, users are encouraged to use that Free Commander XE to verify which files are made available in the VFS. The specific question is:
```
Q: How to see MO's virtual filing system (VFS) in action

A: Use Free Commander or any other suitable windows file manager. For details see 'Free Commander' in the Third Party Programs chapter.
```
This is implying that ALL files, whether inside BSAs or not, are made available for examination. This is in fact what STEP has telling users for some time.
I'm not talking about "magically modifying" files inside BSAs, just the exposure of those files so they can be read, as the game would read them.
Modification of BSA files is not the point of my comment. It is the fact that those BSA files, despite MO saying they win the conflict, are not winning and they remain hidden from tools looking at the VFS.

**Comment by TanninOne** [comment]
a) That's what you're reading into that QA
b) I'm not STEP

The game reads bsas. It has it's own logic to parse bsas. Every tool that works with bsas needs to be able to interpret them. What would be the point of MO replicating that functionality?

You're still confused as to how the BSA handling works: If a file exists inside a BSA and the mod with that bsa has the highest priority among mods that contain this file, MO simply doesn't add the loose file to the vfs at all. The game will then take the file from the bsa because that's the only source.

**Comment by GrantSP** [comment]
I know you're not STEP, that's not the point. STEP is providing the documentation and if we have it wrong it needs to be changed.
The Nexus description and the FAQ I have quoted imply that the VFS that is supplied to the game can be examined, either by eye by viewing it in Free Commander or via a tool, in this case FNIS.
Clearly this is not the case. Both these programmes cannot 'see' what is provided by BSAs and no matter what the user sees by examining the 'Data' tab, that set of files will not be made available if some of those files are in BSAs.

While the description may be true to the game engine, users will, and have been, making the assumption **anything **run via MO will also 'see' the same set of files.

That being the case we need to remove that FAQ and refrain from inviting users to use Free Commander to examine the VFS, as it adds nothing to the user's understanding of the VFS and conflict resolution.
Maybe the description also needs modification, though since MO2 is removing the whole BSA management system, perhaps this entire issue is moot?

**Comment by REALfore** [comment]
Tannin, this doesn't make sense.

Vilja has LOWEST priority, and it get's it's loose file OVERWRITTEN by MO. When I made this test with Vilja and USLEEP both having bsa and loose file, still the lowest priority mod of those gets the loose file overwritten.

Why doesn't MO leave everthing alone, and put the FNIS 0_master.hkx into the overwrite as usual? Why this special bsa logic that a Vilja mod file is overwritten.by FNIS? Sorry, this is NMM strategy, and I don't think it can be the claim of MO.

**Comment by TanninOne** [comment]
@GrantSP  Wait a second. You misinterpret the documentation and as a consequence it needs to be changed? All tools run through MO, that includes Skyrim, fnis and free commander see the same vfs, this is correct information! BSAs are files and they are part of the vfs. The content of BSAs ARE NOT. For no application. never have been. This does not diminish the usefulness of free commander at all.
The only thing that needs to be fixed is your expectation, all documentation you have quoted is correct.

@REALfore If a file from a lower priority mod gets overwritten, *that* is a bug. But this is completely different from what GrantSP described.
The special BSA logic was a mistake and I wanted to remove it for the 1.3.x release but STEP staff convinced me to leave it in because they rely on it. It has been removed in MO2.
However: MO has never and will never place a file that already exists into overwrite. Overwrite is only for new files, existing files that get modified are modified in place. This is necessary for various reasons, among other things an Application could already have a handle open to the existing file.

**Comment by REALfore** [comment]
The why do you ONLY overwrite that (loose) file when there is the same file in the bsa. For the same mod?

FNIS also has an almost vanilla loose 0_master.hkx in its mod. And MO does NOT overwrite it, which is correct and expected. So why is the Vilja 0_master.hkx overwritten instead, just because there is an additional one in the bsa? 

That cannot be justified with handles or that like.

There are situations where I really would LOVE to see that MO would overwrite existing files. Like for PCEA2. But these are files from the overwrite which SHOULD be overwritten. Not the files from a mod, and in particular not a file from an unrelated mod.

**Comment by TanninOne** [comment]
If that's what happens, that is *not* intended. The intended behaviour is always to replace existing files. But, if the application first deletes the file and then creates it from scratch and takes a long time to do so, MO may no longer be able to recognize this as a replacement.

The way bsas play into this is complicated, since i can't delete a file from a bsa but if the file exists in a mod with lower priority, that file is hidden in the vfs and will thus not be replaced (i think).

I'm not denying that this is a mess. As I said, I wanted to get rid of this feature before but the step people needed it.

**Comment by REALfore** [comment]
Why would you EVER replace a file from in mod? bsa or not, the mod authors have their intentions with  the files they provide, and they don't want that overruled by a mod manager.

Maybe there is some situation where it can make sense for pure custom files. But definitely NOT for Skyrim files like 0_master.hkx.

Because what happens when FNIS creates this file anew, and MO uses this file to replace any mod but FNIS itself?
- In case the changed mod does NOT have highest priority of all mods including 0_master.hkx, then most custom animations are ignored. Without any indication what could have gone wrong.
- In case the changed mod (NOT FNIS) does DOES have highest priority, and FNIS is removed, the user will have t-pose without the ability to fix it.

Pretty horrifying scenario from my point of view.

**Comment by TanninOne** [comment]
I've explained this in detail in issue #457. In short: it's technically impossible to reliably get files created in overwrite. If you think you know how then you're missing information.

**Comment by TanninOne** [comment]
> Because what happens when FNIS creates this file anew, and MO uses this file to replace any mod but FNIS itself?

From FNIS perspective you're overwriting a file in skyrim\data without caring what mod the file is from anyway!

> In case the changed mod does NOT have highest priority of all mods including 0_master.hkx, then most custom animations are ignored. Without any indication what could have gone wrong.

You always overwrite the file with the highest priority. This has worked for 6 years and so far this post is the first claim to the contrary.

> In case the changed mod (NOT FNIS) does DOES have highest priority, and FNIS is removed, the user will have t-pose without the ability to fix it.

Like he would with any mod manager or without one.

> Pretty horrifying scenario from my point of view.

Yes, mods can mess up your game, this is how it has always been. MO makes it harder but it can't completely prevent it.

**Comment by REALfore** [comment]
>>> Because what happens when FNIS creates this file anew, and MO uses this file to replace any mod but FNIS itself?
> From FNIS perspective you're overwriting a file in skyrim\data without caring what mod the file is from anyway!

But I overwrite skyrim\data, and NOT the other mod.

>>> In case the changed mod does NOT have highest priority of all mods including 0_master.hkx, then most custom animations are ignored. Without any indication what could have gone wrong.
> You always overwrite the file with the highest priority. This has worked for 6 years and so far this post is the first claim to the contrary.

That's simply not true in several ways. MO has never overwritten 0_master.hkx in the FNIS mod. It only overwrites 0_master,.hkx in Vilja, and only because Vilja has this dual 0_master.
And it doesn't overwrite highest priority (bsa and loose), it overwrites the lowest with bsa and loose.

Maybe I'm the first to report because I'm the only one who has this constallation AND is able to analyze what happens? Why don't YOU test? It's rather easy to reproduce. Grant has done so.

>>> In case the changed mod (NOT FNIS) does DOES have highest priority, and FNIS is removed, the user will have t-pose without the ability to fix it.
>Like he would with any mod manager or without one.

Wrong again.
- 0_master.hkx is only included in the FNIS mod so NMM knows it will have to remove 0_master once FNIS is uninstalled. And t-pose will NOT happen (any more)
- With the uninstall instructions provided by FNIS the user knows what to do. Do you want me to add a note that the MO user has to unistall Vilja as well, becasue MO is unable to handle that? 

>>> Pretty horrifying scenario from my point of view.
> Yes, mods can mess up your game, this is how it has always been. MO makes it harder but it can't completely prevent it.

But it's not that mods mess up the game, it's MO.

I have send a bug report where MO messes up a pretty simple but very awkward scenario (for me as FNIS author). So far I only see you blocking my my reasoning with killer arguments. And you didn't provide ANY verifyable reason why you can't switch off overwriting off loose files (ONLY) when the same file is present in the bsa.

**Comment by GrantSP** [comment]
@Tannin Whilst technically correct, the documentation is misleading and should be modified. Taken with the description that MO manages BSAs in a different manner than other managers, plus the view of the VFS structure as presented by the 'Data' tab it is a very easy conclusion to make that files INSIDE BSAs are also provided to tools by MO. Tools like FNIS and Free Commander.
I can see now this is incorrect and sadly I have been under a misunderstanding for some time about this, that's my error not the code of MO. However, using the documentation as a guide provided no solid indicator to the contrary. I will rectify that situation.

@REALfore 
As noted in the PM to you, there is no issue with the modifying of the _0_master.hkx_ in this matter.
It is all due to priority of the mods, which in a manual installation is the same as the install order.

When I installed FNIS & Vilja manually into Skyrim, without using MO at all, the same thing happens.
You run the tool and the file that was supplied by either FNIS or Vilja was modified, how is this any different to what is managed by MO? If I installed Vilja after FNIS, the copy from that mod is now in **Skyrim\Data\** and that file is used. If I install FNIS after Vilja then the FNIS version is used.
Regardless of whether MO is used or not, FNIS modifies the _0_master.hkx_ that is in **Skyrim\Data\** .

**Comment by REALfore** [comment]
Grant, please read and digest my previous posts before telling me there is no issue.

First, there is no sense at all telling that Vilja has to be installed AFTER Vilja. This is a totally unnecessary dependency (only because MO has a bug)

Second, MO modifies Vilja's MOD structure. And when FNIS is uninstalled, the modified 0_master.hkx (in Vilja) will cause all kind of animation issues, including general t-pose. Meaning the t-pose will exist until Vilja is REMOVED from the MO mod list. That has NOTHING to do with improper uninstall for FNIS. 

And btw. there will be no character t-pose when using NMM and uninstallinf FNIS.

**Comment by TanninOne** [comment]
@GrantSP I do realize this whole feature is confusing. When we discussed this in the step-thread it took around 3 pages of discussion and explanation until any of the step-people themselves understood how MOs bsa handling actually works and what the consequences are. I've come to hate this feature and wish I had never put it in.
Anyway: MOs vfs is still simply a filesystem-level tool, it only manipulates file lookup, never file content (Except for ini-files, another mess I've dropped in MO2).
So bsas are still simply files, the bsa management works by hiding loose files, not by making files inside bsas visible as regular files.
The data-view in the UI shows bsa content, but this is a UI feature and files from bsas are displayed in italics to signify they are "different".
The data-view is primarily there to see file-conflicts, not to give a representation of how the vfs will look like. That's part of the reason I have been advertising Free Commander as a tool to see the actual vfs.

@REALfore could you please calm down a little? Your posts come across as very aggressive and I have very little wish to continue this discussion in this tone.

Still, here are a few responses:

a) But I overwrite skyrim\data, and NOT the other mod.
You overwrite files in skyrim\data without knowing which mod provided them. They could be from fnis, they could be from any other mod installed after fnis.
If you use nmm to install fnis, then vilja, the 0_master.hkx in skyrim\data will be from vilja. If fnis changes it, it will change the file provided by vilja. If you uninstall fnis, to my knowledge, it should leave the modified 0_master.hkx in place. Barring any bugs, MO should provide the same situation as the unmodifie hkx from Vilja no longer exists in both cases.

b) That's simply not true in several ways. MO has never overwritten 0_master.hkx in the FNIS mod.
This is incorrect, MO will overwrite the 0_master.hkx from fnis if that is the highest priority mod providing it and that's a fact. I've been using MO with fnis like this for years.
I grant that it's possible there may be a bug if a bsa with the same file involved, but as long as only loose file are concerned it's definitively the highest priority file being overwritten.

c) Why don't YOU test?
Because I currently have very little spare time and the time I have I spend on MO2. There is no way I'm going to spend time investigating a bug in MO1 that hasn't bothered anyone in several years when MO2 users are still waiting for a non-alpha release.

d) Grant has done so.
Except that GrantSP has been reporting something different and you just flipped out on him when he posted that.

e) So far I only see you blocking my my reasoning with killer arguments.
That's bullsh*t. My only responses to your bug report have been "If a file from a lower priority mod gets overwritten, that is a bug." and "If that's what happens, that is not intended.".
I've at no point blocked your bug report, what I have been refuting is your demand MO shouldn't be modifing files in mods at all. And not because I don't want to but because I've been considering this for a long time and it simply can't be done without driver-level functionality. Or if it can be done, I'm not smart enough.

**Comment by REALfore** [comment]
I don't call it aggressive, but certaily irritated. How many hours have I spent informing you about an MO bug, which can in fact hamper many users. Simply because Vilja, MO and FNIS have this high number of users. And instead of seeing any support in dealing with this issue I only get counter-arguments which totally ignore what we sould care fore: our users. So again:

a) But I overwrite skyrim\data, and NOT the other mod.
I DO know what I do overwrite. Because all these 0_master.hkx (Vilja, USLEEP) were created by myself to fix the so-called hug-after-brawl-bug for all those users who don't use FNIS otherwise.

b) That's simply not true in several ways. MO has never overwritten 0_master.hkx in the FNIS mod.

> This is incorrect, MO will overwrite the 0_master.hkx from fnis if that is the highest priority mod providing it and that's a fact. I've been using MO with fnis like this for years.

This is the part which baffles me most. Isn't the overwrite your very own concept? 0_master.hkx is ALWAYS put into overwrite, not into the FNIS mod. I have my new PC now for a month, and probably have run FNIS 100 times ever since. And in my FNIS mod there is still this original file, dated 9-13-2014. THe overwrite is up-to-date (now that I removed the Vilja loose file). Are we talking about different things?

c) d) e)
Sorry, but I also have VERY little time. Actually I'm not modding any more, and I'm only supporting my mods. But here I see an issue which can explain many unsolved animation mysteries. And all we do here is getting torn up in discussions which are needed like a hole in the head.

What I see is that MO does something special and WRONG in a certain situation. Overwriting a file in case that the same file is in bsa and loose. All I say that this special treatment is a bug. And don't tell me it takes driver-level functionality to switch this needless treatment off.

**Comment by GrantSP** [comment]
I've figured it out!!

@REALfore The setup you are running is different to that which I, and most users are running.
Let me explain.
When I install FNIS the meshes folder containing _0_master.hkx_ is also installed and it is part of the Skyrim Data structure. Running FNIS like this DOES modify that file (if this mod is in the greatest priority).
Now I have just renamed ALL instances of 0_master.hkx found in my installation, thus there are NO copies of this file in Skyrim's Data (in fact no copies anywhere on my system), and run FNIS and what do I find: a NEW file called _0_master.hkx_ is created in overwrite.
This is expected as MO has always created NEW files into 'Overwrite' but modified EXISTING files in the location they where found.
So either the folder where FNIS has its meshes is NOT part of the actual installation OR you still have files in your overwrite (files here always have the greatest priority) and they were being modified when you were using FNIS.
This is the only way I can replicate the behaviour you describe.

**Comment by TanninOne** [comment]
> How many hours have I spent informing you about an MO bug, which can in fact hamper many users. 

I don't know, how many hours? This is the first bug report I got on the topic and it's over a year after I announced I would do only the absolutely necessary on the old vfs lib and you attack me because I don't provide a solution yet?
All three involved mods have been around for several years. If you have been aware of this problem and MOs possible involvement you could have let me in on it at a time where I was still actively working on the component, that would have been a hoot...

> This is the part which baffles me most. Isn't the overwrite your very own concept?

Yes, and the concept of overwrite was **always** only a fallback when MO can't determine where else to place new files. It was never intended as a catch-all for file modifications. I never advertised it as such. I corrected at least a dozen forum posts where it was described as such.

> THe overwrite is up-to-date (now that I removed the Vilja loose file). Are we talking about different things?

No, this is exactly what I've been explaining before: MO replaces the highest priority file, overwrite is always the highest priority. If you have a 0_master.hkx in overwrite, THAT file will be overwritten. If you remove it, MO will do it's best to replace the existing file.

> And all we do here is getting torn up in discussions which are needed like a hole in the head.

Then why are you turning this into a discussion of principles with statements like "Why would you EVER replace a file from in mod?"
Don't act like it wasn't you who sidetracked the issue.

> And don't tell me it takes driver-level functionality to switch this needless treatment off.

I didn't, please re-read my post. I made that statement regarding driver-level support in a completely different context.

**Comment by TanninOne** [comment]
So I now got around to testing this and I can't reproduce the described behaviour.

I installed fnis 6.2, current vilja, all other mods deactivated, empty overwrite.
Fnis had higher priority. The file "meshes\actors\character\behaviors\0_master.hkx" now exists in fnis, in vilja as a loose file and in vilja as a bsa.
I then ran GenerateFNISForUsers.exe. The only file touched was the one in fnis.
Then I reset everything, deleted the loose file file in vilja, ran again. The only file touched was the one in fnis.
Then I reset everything, this time deleted the loose file in fnis, ran again. This time, the loose file in vilja was updated.
Then I reset everything, this time deleted both loose files, ran again. This time a new 0_master.hkx was created in overwrite.

So from my testing, everything works as intended.

**Comment by REALfore** [comment]
I did exactly the same you did.

> I installed fnis 6.2, current vilja, all other mods deactivated, empty overwrite.
Fnis had higher priority. The file "meshes\actors\character\behaviors\0_master.hkx" now exists in fnis, in vilja as a loose file and in vilja as a bsa.
I then ran GenerateFNISForUsers.exe. The only file touched was the one in fnis.

I got an FNIS Unhandled exception 
The file C:\Games\Steam\steamapps\Skyrim\data\meshes\actor\character\behaviors\1hm_behavior.hkx could not befound

This is a Win10 related issue with double "\\" in the path on certain OS functions. Since you didn't get it I assume you don't run Win10.

I updated with a modified FNIS generator. This time everything works. Except:
Vilja's loose 0_master.hkx is overwritten.

'Then I reset everything, deleted the loose file file in vilja, ran again. The only file touched was the one in fnis.

This time a new 0_master.hkx was created in overwrite.

In addition 2 other files are written into Skyrim/data (not Overwrite, not FNIS mod). For these files I also see ERRORs in the MO log:
ERROR (00:12:53.0266): not a mod directory: C:\GAMES\Steam\steamapps\common\Skyrim\data\meshes\animationsetdatasinglefile.SAVE
ERROR (00:12:53.0285): not a mod directory: C:\GAMES\Steam\steamapps\common\Skyrim\data\meshes\animationsetdatasinglefile.txt

I had this errors before when I reported my problems to GrantSP. However they disappeared after I moved Steam and MO away from C:\Program (x86), and granted all possible rights to both trees. But now they are back again after I set up MO in the way you have described (Installed Vilja and FNIS anew, deactivated everything else)

'Then I reset everything, this time deleted the loose file in fnis, ran again. This time, the loose file in vilja was updated.
Then I reset everything, this time deleted both loose files, ran again. This time a new 0_master.hkx was created in overwrite.

Those 2 tests I didn't make, since it's uncontested.

**Comment by GrantSP** [comment]
I can confirm the unhandled exception errors as described here. They are exactly as I posted i above with regards to these two mods.

Further, if I rename the loose file from those two mods and I create a dummy mod that contains only _0_master.hkx_ and the order of that mod is anywhere above (ie. a lower priority) the Vilja mod (with the BSA0 it errors out.
Order this mod anywhere below (with a higher priority) the mod containing the BSA and the tool runs and creates a new file.

For some reason the Vilja mod BSA's position has a bearing on whether FNIS looks for an existing file to work on. It apparently sees it but can't access it.

**Comment by REALfore** [comment]
> I can confirm the unhandled exception errors as described here. They are exactly as I posted i above with regards to these two mods.

I forgot to mention: this unhandled exception does NOT happen when FNIS is run loose (not under MO)

**Comment by REALfore** [comment]
Now MO messes things up so badly, it's terrible.

In the FNIS generator I get an unhandled exception "Index out of bounds" INSIDE DeleteFile().
Reason: because of my test with FNIS loose files there was a 1hm_behavior.hkx remnant in the Skyrim/data tree.

Unhandled error(9) have been reported so often by MO users before, and I never knew the reason. Now I do.

**Comment by GrantSP** [comment]
> Reason: because of my test with FNIS loose files there was a 1hm_behavior.hkx remnant in the Skyrim/data tree.
> 
> Unhandled error(9) have been reported so often by MO users before, and I never knew the reason. Now I do.

Which is precisely why instructions are given to have a clean "vanilla" Data folder to work from.
You initial response to this when I told you was:

> Now Grant, this is TOTAL nonsense.
Now you know it is not nonsense.

**Comment by TanninOne** [comment]
Please provide logs for your 2 testruns.
And just a few things
a) I also run windows 10
b) MOs vfs lib is a native C/C++ library. It cannot be the source of your indexoutofbounds exception.
c) I also analyzed a couple of cases of file not found exceptions and they have always been caused by fnis deleting a file, then trying to access it later without re-creating them in between.

Look, if you need help bugfixing the generator, you could just publish current source code. Maybe someone can help.
But blaming bystanders is a bit silly. MOs vfs works with every application thrown at it so far, including monsters like the java vm and the python interpreter. But your tool exposed all these wonderous problems like index out of bounds exceptions from an api that does no boundary checking...

**Comment by REALfore** [comment]
> b) MOs vfs lib is a native C/C++ library. It cannot be the source of your indexoutofbounds exception.
c) I also analyzed a couple of cases of file not found exceptions and they have always been caused by fnis deleting a file, then trying to access it later without re-creating them in between.

This, Tannin, is the perfect example why we are unable to deal with one another, and why I am so pissed. There are bugs in MO, and you constantly blame FNIS implementation or user behavior.

"It cannot be the source of your indexoutofbounds"? Then why doesn't it ever happen for non-MO users? And how can it possibly be caused by a simple DeleteFile("string_much_shorter_than_240_chars")?

"I also analyzed a couple of cases"? Yes, Grant has showed me one of your posts. A perfect example of an MO BUG. There are numerous cases where OS functions return inconsistent results. A fileExist() or GetFiles() shows the existance of a file, a subsequent DeleteFile() or GetVersionInfo() fails. Always MO. Exactly the case in that post.

The same for this Process.WaitForExit() which doesn't properly work with MO involved. I have HEAVILY restructured FNIS (Grant knows) in order to give MO time to update its data. But I can't work aroud each and every MO bug.

> Now you know it is not nonsense.

Yes Grant, it IS nonsense. You can't keep users from installing stuff manually, and then blame them for having done so. And who has to deal with the consequences? Not tannin, you only to a little extent. No, IT IS ME (and the user), and I am sick and tired of it. Since more than 3 years.


Unless you, tannin, are actively going to do something, and take active lead, I will stop publicly any support related to MO.

**Comment by TanninOne** [comment]
> This, Tannin, is the perfect example why we are unable to deal with one another, and why I am so pissed. There are bugs in MO, and you constantly blame FNIS implementation or user behavior.

I think it's more likely you're so pissed off because your hormones are acting up....
Seriously, you have to be joking. This is a post in MOs issue tracker, to my knowledge you don't even have one. Your posts include "MO will overwrite the mod with lower priority", "Tannin, this doesn't make sense.", "Why doesn't MO leave everthing alone", "The why do you ONLY overwrite that (loose) file when there is the same file in the bsa.", "Why would you EVER replace a file from in mod?", "That's simply not true in several ways.", "Pretty horrifying scenario from my point of view.", "But it's not that mods mess up the game, it's MO.", "Now MO messes things up so badly, it's terrible."

Every post you make is aggressive, unobjective and all about laying blame. And when I try to add facts to the table, like what is/isn't achievable with a user-level hooking lib and how a native application can't throw a managed exception (which is easy to verify, you don't have to take my word for it), *you* accuse *me* of blocking and trying to blame fnis.
This is ridiculous and I'm no longer playing. This is the last post from you I will answer to until you get your temper under control. If anyone else is willing to seriously try to figure this problem I will happily continue to investigate but another aggressive post from you fore and I will ban you from my github-projects. I don't have the nerve for this.

> Then why doesn't it ever happen for non-MO users?

Correlation does not equal causation! This is a common misconception but fatal for a software developer.
With the same logic I could ask: Why are these supposed MO problem only occuring in fnis and not in tesxedit, skproc, skyrim, fallout x, fomm, cmd, free commander,  loot/boss, ... ?

>And how can it possibly be caused by a simple DeleteFile("string_much_shorter_than_240_chars")?

It can't, or better: it couldn't if you were referring to the native DeleteFile call MO modifies. But your code is managed vb.net code! There are at least 1-2 abstraction layers between the library you use and the api MO affects and I have no way of knowing what happens between.

> There are numerous cases where OS functions return inconsistent results. A fileExist() or GetFiles() shows the existance of a file, a subsequent DeleteFile() or GetVersionInfo() fails.

Neither fileExist nor GetFiles are OS functions, those are library functions. And there is no 1:1 equivalent in the native api, whereas DeleteFile and GetVersionInfo are probably simple proxies to the equivalent native api with parameter/result conversion from native<->managed.
So my guess is that the former two functions may accept and transform parameters that are otherwise not valid in the native api. (and before you reply to that: "valid in the native api" and "usually accepted by the native api" are not the same! Your code can be wrong and still produce correct results most of the time.)
But, since you don't provide code and I don't know what logs you refer to all I can do is guess.

> The same for this Process.WaitForExit() which doesn't properly work with MO involved.

MO doesn't affect any wait functions so this is clearly a red herring.

>  I have HEAVILY restructured FNIS (Grant knows) in order to give MO time to update its data

yeah? well, maybe you should have talked to me instead? I had an issue tracker the whole time and the last time you contacted me directly you asked me for a feature in MO and I implemented it in the next release so what is this BS "we are unable to deal with one another" about? You never talked to me about these things and now it's my fault you hacked around them instead of solving them?

> Unless you, tannin, are actively going to do something, and take active lead, I will stop publicly any support related to MO.

I am! I'm actively asking the only person I know of who can reproduce the issue to provide logs and source code but all he does is insult and now threaten.

**Comment by FrostByghte** [comment]
While I realize this is for bug regarding MO, this topic is also being discussed here on Nexus:

https://forums.nexusmods.com/index.php?/topic/822684-vilja-in-skyrim/page-3195#entry36617475

http://www.nexusmods.com/skyrim/mods/26393/ (comment section)

Emma now has a warning stickied at the top of the Vilja mod indicating issues with MO that could cause problems.  Seriously anyone that reads this bug report can see there is a user having problems with MO. The statements being made both here and on the nexus have seriously miss-represented the issue at hand and cast MO in an unfavorable manner that is not valid.

**Comment by TanninOne** [comment]
I'm closing this issue as it has been derailed completely. The primary issue that can be attributed to MO is discussed further in issue #654
